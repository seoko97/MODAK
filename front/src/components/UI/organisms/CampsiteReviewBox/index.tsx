import React, { useCallback, useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import ClipLoader from "react-spinners/ClipLoader";

import Tab from "@molecules/ReviewCard/Tap";
import SortedReview from "@molecules/SortedReview";
import ReviewCard from "@molecules/ReviewCard";
import useScroll from "@hooks/useScroll";
import { useAppSelector, AppDispatch } from "@store/configureStore";
import useThrottle from "@hooks/useThrottle";
import { getCampReviews } from "@reducers/reviews/action";

const StyledContainer = styled.section`
  width: 100%;
  min-height: 1000px;
  & * {
    color: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};
  }
`;

const NoneReview = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 0.7;
  margin-top: 40px;
`;
const BoxHeader = styled.div`
  position: sticky;
  top: 72px;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background-color: ${({ theme }) => theme.BAKCGROUND_COLOR.RGBA};
  backdrop-filter: saturate(180%) blur(20px);
  z-index: 10;
`;

const CampsiteReviewBox = () => {
  const dispatch: AppDispatch = useDispatch();
  const { mainReviews, getCampReviews: CRState } = useAppSelector((state) => state.reviews);
  const { create } = useAppSelector((state) => state.review);
  const { singleCamp } = useAppSelector((state) => state.camp);
  const [currTab, setCurrTab] = useState("전체");
  const [skip, setSkip] = useState(0);
  const [scrollHeight, clientHeight] = useScroll();
  const [sorted, setSorted] = useState("_id");

  const checkedRating = useCallback((tab: string) => (tab === "전체" ? "" : tab), []);

  const onThrolttle = useThrottle(async () => {
    await dispatch(
      getCampReviews({
        campId: singleCamp?._id as string,
        skip: String(skip + 1),
        rating: checkedRating(currTab),
        target: sorted,
      }),
    );
    setSkip(skip + 1);
  }, 500);

  const handleClickTab = useCallback(
    (tab: string) => {
      setSkip(0);
      setCurrTab(tab);

      dispatch(
        getCampReviews({
          campId: singleCamp?._id as string,
          rating: checkedRating(tab),
          target: sorted,
        }),
      );
    },
    [singleCamp?._id, sorted],
  );

  useEffect(() => {
    if (scrollHeight + 300 >= clientHeight) {
      onThrolttle();
    }
  }, [scrollHeight, clientHeight]);

  useEffect(() => {
    if (create.done) setSorted("_id");
  }, [create.done]);

  const onChange = useCallback(async (e) => {
    const query = {
      target: e.target.value,
    };
    await dispatch(getCampReviews({ campId: singleCamp?._id as string, ...query }));
    setSorted(e.target.value);
  }, []);

  return (
    <StyledContainer>
      <BoxHeader>
        <SortedReview onChange={onChange} />
        <Tab current={currTab} onClick={handleClickTab} />
      </BoxHeader>
      {mainReviews[0] ? (
        mainReviews.map((review) => <ReviewCard key={review._id} review={review} />)
      ) : (
        <>
          {CRState.loading ? (
            <NoneReview>
              <ClipLoader color="#ccc" />
            </NoneReview>
          ) : (
            <NoneReview>리뷰가 존재하지 않습니다.</NoneReview>
          )}
        </>
      )}
    </StyledContainer>
  );
};

export default CampsiteReviewBox;
