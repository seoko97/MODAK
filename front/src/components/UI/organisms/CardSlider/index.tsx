import React, { memo, useMemo } from "react";
import styled from "styled-components";
import Slider, { Settings } from "react-slick";
import TitleBox from "@molecules/TitleBox";

import { useAppSelector } from "@store/configureStore";
import { NextArrow, PrevArrow } from "./arrow";

const Container = styled.article`
  padding: 20px 0;
  width: 100%;
  & .slick-arrow {
    width: 14px;
    display: block;
    &.reverse {
      transform: rotate(180deg) translateY(50%);
    }
  }
  & .post-slide {
    width: 98%;
    margin: 0px auto;
    padding: 20px 0;
  }
  & .slick-slide {
    padding: 10px 0;
  }

  & svg {
    fill: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};
  }

  @media (max-width: ${({ theme }) => theme.BP.HDPC}) {
    & .post-slide {
      width: 95%;
    }
  }

  @media (max-width: 600px) {
    & .post-slide {
      width: 100%;
      margin: 0;
    }
  }
`;

const settings: Settings = {
  infinite: true,
  slidesToScroll: 1,
  speed: 500,
  nextArrow: <PrevArrow />,
  prevArrow: <NextArrow />,
  autoplaySpeed: 4000,

  responsive: [
    {
      breakpoint: 980,
      settings: {
        slidesToShow: 2,
      },
    },
    {
      breakpoint: 600,
      settings: {
        slidesToShow: 1,
        arrows: false,
      },
    },
  ],
};

interface Props {
  title: string;
  children: React.ReactNode;
}

const CampCardSlider = ({ children, title }: Props) => {
  const { mainCamps } = useAppSelector((state) => state.camps);

  const showLength = useMemo(() => {
    const postsLength = mainCamps.length;
    return postsLength < 3 ? postsLength : 3;
  }, [mainCamps]);

  return (
    <Container>
      <TitleBox title={title} />
      <div className="post-slide">
        <Slider {...settings} slidesToShow={showLength}>
          {children}
        </Slider>
      </div>
    </Container>
  );
};

export default memo(CampCardSlider);
