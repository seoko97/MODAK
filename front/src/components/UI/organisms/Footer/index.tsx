import React from "react";
import styled from "styled-components";
import RowFrame from "@templates/RowFrame";
import FooterContent from "./FooterContent";

const Wrapper = styled.footer`
  position: absolute;
  bottom: 0;
  width: 100%;
  height: 140px;
  background-color: #3e3e3e;
  color: #9b9b9b;
  font-size: 16px;
  & > div {
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: space-around;

    & svg {
      width: 140px;
    }
    .fill-logo {
      fill: #fff;
    }
  }

  @media (max-width: ${({ theme }) => theme.BP.MOBILE}) {
    & > div {
      flex-direction: column;
    }
  }
`;

const Footer = () => {
  return (
    <>
      <Wrapper>
        <RowFrame>
          <FooterContent />
        </RowFrame>
      </Wrapper>
    </>
  );
};

export default Footer;
