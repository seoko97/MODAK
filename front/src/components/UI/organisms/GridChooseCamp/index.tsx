import React from "react";
import styled from "styled-components";
import GridCampInfo from "@molecules/GridCampInfo";
import TitleBox from "@molecules/TitleBox";

const Container = styled.div`
  padding: 20px 0;
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 20px;
  margin-bottom: 20px;
`;

const StyledCamp = styled.div`
  width: 100%;
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  gap: 20px;

  & > div {
    background-color: #ccc;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    grid-template-columns: repeat(1, 1fr);
    gap: 15px;
  }
`;

const GridChooseCamp = () => {
  return (
    <Container>
      <TitleBox title="이런 캠핑장은 어떠세요?" />

      <StyledCamp>
        <GridCampInfo
          title="#봄꽃여행"
          description=""
          to="/search?thema=봄꽃여행"
          src="/spring.jpg"
        />
        <GridCampInfo
          title="#여름물놀이"
          description=""
          to="/search?thema=여름물놀이"
          src="/summer.jpg"
        />
        <GridCampInfo
          title="#가을단풍명소"
          description=""
          to="/search?thema=가을단풍명소"
          src="/autumn.jpg"
        />
        <GridCampInfo
          title="#겨울눈꽃명소"
          description=""
          to="/search?thema=겨울눈꽃명소"
          src="/winter.jpg"
        />
      </StyledCamp>
    </Container>
  );
};

export default GridChooseCamp;
