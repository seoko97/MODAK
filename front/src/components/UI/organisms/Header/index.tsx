import React from "react";

import HeaderSection from "./HeaderSection";
import MobileHeaderSection from "./MobileHeaderSection";

const Header = () => {
  return (
    <>
      <HeaderSection />
      <MobileHeaderSection />
    </>
  );
};

export default React.memo(Header);
