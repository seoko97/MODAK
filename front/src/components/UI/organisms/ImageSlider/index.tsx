import React, { useCallback, useRef } from "react";

import Slick from "react-slick";

import { url } from "@apis/.";
import { Inner, MainSlickItems, NextButton, NextIcon, PrevButton, PrevIcon, Wrap } from "./styles";

interface Props {
  images: string[];
}
const setting = {
  dots: true,
  arrows: false,
  infinite: true,
  slidesToShow: 1,
  slidesToScroll: 1,
};

const Slide = ({ images }: Props) => {
  const mainSlickRef = useRef(null);

  const onClickPrev = useCallback((ref) => () => ref.current.slickPrev(), []);
  const onClickNext = useCallback((ref) => () => ref.current.slickNext(), []);
  const len = images.length;
  return (
    <Wrap>
      <Inner>
        <Slick ref={mainSlickRef} {...setting}>
          {images.map((v, i) => {
            return (
              <MainSlickItems key={`${v}_${i}`}>
                <img src={`${url}/${v}`} alt="img" />
              </MainSlickItems>
            );
          })}
        </Slick>
        <>
          {len !== 1 && (
            <>
              <PrevButton onClick={onClickPrev(mainSlickRef)}>
                <PrevIcon />
              </PrevButton>
              <NextButton onClick={onClickNext(mainSlickRef)}>
                <NextIcon />
              </NextButton>
            </>
          )}
        </>
      </Inner>
    </Wrap>
  );
};
export default Slide;
