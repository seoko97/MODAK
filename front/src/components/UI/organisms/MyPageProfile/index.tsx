import React from "react";
import useModal from "@src/hooks/useModal";
import { useAppSelector } from "@src/store/configureStore";
import { EditUserData } from "@src/types/apis/user";
import ProfileImage from "./ProfileImage";
import ProfileInfo from "./ProfileInfo";
import Style from "./styles";

const MyPageProfile = (): React.ReactElement => {
  const { me, userInfo } = useAppSelector((state) => state.user);
  const [isOpen, onOpen, onClose] = useModal();
  const { nickname, profileImg, intro } = me as EditUserData;
  const displayEdit = me?._id === userInfo?._id;

  return (
    <Style.Profile>
      <ProfileImage profileImg={profileImg} />
      <ProfileInfo
        displayEditButton={displayEdit}
        nickname={nickname}
        onOpen={onOpen}
        isOpen={isOpen}
        onClose={onClose}
        intro={intro}
      />
    </Style.Profile>
  );
};

export default MyPageProfile;
