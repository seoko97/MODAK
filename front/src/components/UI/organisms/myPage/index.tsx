import styled from "styled-components";
import Link from "next/link";
import React, { memo, ReactElement, useState } from "react";
import SubTitle from "@src/components/UI/atoms/SubTitle";
import HeartIcon from "@src/components/icons/HeartIcon";
import PencilIcon from "@src/components/icons/PencilIcon";
import MyPageProfile from "@src/components/UI/molecules/MypageProfile";
import Title from "@src/components/UI/atoms/Title";
import Tabs from "@src/components/UI/organisms/Tabs";

export type TabList = {
  "내 리뷰": ReactElement;
  "나의 캠핑 기록": ReactElement;
  "찜한 캠핑장": ReactElement;
};

const IconComponent = () => (
  <IconBox>
    <Icons>
      <PencilIcon size={13} /> 10
    </Icons>
    <Icons>
      <HeartIcon size={13} /> 10
    </Icons>
  </IconBox>
);

const HeartIconComponent = () => (
  <IconBox>
    <Icons>
      <HeartIcon size={13} /> 10
    </Icons>
  </IconBox>
);

const MyPage = () => {
  const review = {
    _id: 1,
    content: `안녕하세요! 저번 주말 선재도의 트리 캠핑장에 다녀왔습니다.
      여기는 경기도 지역 오션뷰 캠핑장으로 정말 유명한 곳인데요.
      홈페이지에서 예약할 때 정보가 별로 없다보니 어떤자리를 예약하는게 좋을지 분간이 잘 안되서 영상을 직접 찍게 되었습니다.
      
      이 곳은 정말 숲 속에 있는 분위기에 바다까지 볼 수 있어서 낭만적인 분위기를 연출할 수 있는 곳입니다.
      또한 겨울에 와서 그런지 벌레도 전혀 없고 한적해서 정말 좋았어요.
      시도 때도 없이 날아다니는 비행기 소리가 조금 거슬리긴 하지만..
      인천지역 트리 캠핑장 추천합니다!
      
      =======================================================================================================
      E=MAIL : 88leeq@naver.com
      인스타그램 : https://www.instagram.com/88_leeq`,
    photos: [
      "https://images.unsplash.com/photo-1594495894542-a46cc73e081a?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1171&q=80",
      "https://images.unsplash.com/photo-1563299796-17596ed6b017?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80",
      "https://images.unsplash.com/photo-1475483768296-6163e08872a1?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80",
    ],
    author: {
      _id: 2,
      nickname: "김불멍",
      profile:
        "https://images.unsplash.com/photo-1531427186611-ecfd6d936c79?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80",
      likes: 123,
      posts: 13,
    },
    createAt: "2022-01-28",
    bookmarks: 173,
  };

  const { _id, content, photos, author, createAt, bookmarks } = review;
  const { nickname, profile, likes, posts } = author;

  const ReviewComponent = () => (
    <Review>
      <CampInfo>
        <SubTitle size={14}>{createAt}</SubTitle>
        <CampLink href="#">
          <a>@장호비치캠핑장</a>
        </CampLink>
        <HeartIconComponent />
      </CampInfo>
      <ReviewPost>
        안녕하세요! 저번 주말 선재도의 트리 캠핑장에 다녀왔습니다.<br></br>
        여기는 경기도 지역 오션뷰 캠핑장으로 정말 유명한 곳인데요.<br></br>
        홈페이지에서 예약할 때 정보가 별로 없다보니 어떤자리를 예약하는게 좋을지 분간이 잘 안되서
        영상을 직접 찍게 되었습니다.<br></br>
        <br></br>이 곳은 정말 숲 속에 있는 분위기에 바다까지 볼 수 있어서 낭만적인 분위기를 연출할
        수 있는 곳입니다.<br></br>
        또한 겨울에 와서 그런지 벌레도 전혀 없고 한적해서 정말 좋았어요.<br></br>
        시도 때도 없이 날아다니는 비행기 소리가 조금 거슬리긴 하지만..<br></br>
        인천지역 트리 캠핑장 추천합니다!<br></br>
        <br></br>
        =======================================================================================================
        <br></br>
        E=MAIL : 88leeq@naver.com<br></br>
        인스타그램 : https://www.instagram.com/88_leeq
      </ReviewPost>
      <PhotoBox>
        {photos.map((photo, idx) => (
          <img key={idx} src={photo} alt="reviewPhoto" />
        ))}
      </PhotoBox>
    </Review>
  );

  const VisitedComponent = () => (
    <VisitedCamp>
      <SubTitle>2022-01-28</SubTitle>
      <CampInfo>
        <CampLink href="#">
          <a>@캠핑장 이름/ 주소</a>
        </CampLink>
        <IconComponent />
      </CampInfo>
    </VisitedCamp>
  );

  const WishComponent = () => (
    <WishCamp>
      <CampInfo>
        <CampLink href="#">
          <a>
            <p>@캠핑장 이름/ 주소</p>
          </a>
        </CampLink>
        <IconComponent />
      </CampInfo>
      <Tags>#가족캠핑 #글램핑</Tags>
      <OtherReview>
        <Title size={14}>주요 리뷰</Title>
        <PhotoBox>
          {photos.map((photo) => (
            <OtherLink href="#" key={1}>
              <img src={photo} alt="reviewPhoto" />
            </OtherLink>
          ))}
        </PhotoBox>
      </OtherReview>
    </WishCamp>
  );

  const tabList: TabList = {
    "내 리뷰": <ReviewComponent />,
    "나의 캠핑 기록": <VisitedComponent />,
    "찜한 캠핑장": <WishComponent />,
  };

  const [tab, setTab] = useState<keyof TabList>("내 리뷰");

  const handleClick = (tab: keyof TabList) => {
    setTab(tab);
  };

  return (
    <Main>
      <MyPageProfile />
      <Tabs current={tab} onClick={handleClick}></Tabs>
      {tabList[tab]}
    </Main>
  );
};

const IconBox = styled.div`
  display: flex;
`;

const Icons = styled.div`
  display: flex;
  box-sizing: border-box;
  padding: 5px;
  gap: 6px;
  justify-content: center;
  align-items: baseline;
  font-size: 14px;
  cursor: pointer;
  :hover {
    font-weight: bold;
  }
  @media (max-width: ${({ theme }) => theme.BP.MOBILE}) {
    padding: 0;
  }
`;

const CampInfo = styled.div`
  display: flex;
  gap: 10px;
  align-items: baseline;
  margin-bottom: 14px;
  & a {
    color: #038c5a;
  }
  @media (max-width: ${({ theme }) => theme.BP.MOBILE}) {
    flex-direction: column;
    gap: 2px;
  }
`;

const CampLink = styled(Link)`
  display: flex;
  cursor: pointer;
`;

const PhotoBox = styled.div`
  display: flex;
  margin-top: 60px;
  & img {
    width: 33.3%;
    @media (max-width: ${({ theme }) => theme.BP.MOBILE}) {
      width: 100%;
      + img {
        display: none;
      }
    }
  }
`;

// --- 전체 컨테이너 ---
const Main = styled.div`
  padding: 20px;
`;

const Review = styled.section`
  padding-left: 10px;
  padding-right: 10px;
  & p {
    margin-bottom: 10px;
  }
`;

const ReviewPost = styled.div`
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 9;
  -webkit-box-orient: vertical;
  word-wrap: break-word;
  line-height: 160%;
`;

const VisitedCamp = styled.div``;

const WishCamp = styled.div``;

const Tags = styled.div`
  display: flex;
`;

const OtherReview = styled.div``;

const OtherLink = styled(Link)``;

export default memo(MyPage);
