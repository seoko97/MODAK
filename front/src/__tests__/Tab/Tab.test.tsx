import { TabList } from "@src/components/UI/organisms/MyPage";
import React from "react";
import renderer from "react-test-renderer";
import Tab from "../../components/UI/organisms/Tab";

describe("Tab 테스트", () => {
  const tabs: (keyof TabList)[] = ["내 리뷰", "나의 캠핑 기록", "찜한 캠핑장"];
  let current: keyof TabList;
  let onClick: () => void;

  beforeEach(() => {
    current = "내 리뷰";
    onClick = jest.fn();
  });

  it("렌더 확인", () => {
    const component = renderer.create(<Tab current={current} onClick={onClick} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
