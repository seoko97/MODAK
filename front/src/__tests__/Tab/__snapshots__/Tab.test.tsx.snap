// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab 테스트 렌더 확인 1`] = `
<div
  className="sc-bdvvtL jZanzH"
>
  <span
    className="sc-gsDKAQ bITfVn"
    onClick={[Function]}
  >
    내 리뷰
  </span>
  <span
    className="sc-gsDKAQ hFbhMX"
    onClick={[Function]}
  >
    나의 캠핑 기록
  </span>
  <span
    className="sc-gsDKAQ hFbhMX"
    onClick={[Function]}
  >
    찜한 캠핑장
  </span>
</div>
`;
